Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,746 @@
+```
+
+Example 6:
+Using multidimensional slices in copy-and-update expressions.
+
+```qsharp
+let zeros =[|0, size = (3, 3)|];
+
+let withCorners = zeros w/ (0..2..2, 0..2..2) <- [|1, 2|, |3, 4|];
+// withCorners = [|1, 0, 2|, |0, 0, 0|, |3, 0, 4|] and is of type [|Int|].
+
+let withFirstRow = zeros w/ (0, 0..2) <- [1, 2, 3];
+// withFirstRow = [|1, 2, 3|, |0, 0, 0|, |0, 0, 0|] and is of type [|Int|].
+
+let withFirstColumn = zeros w/ (0..2, 0) <- [1, 2, 3];
+// withFirstColumn = [|1, 0, 0|, |2, 0, 0|, |3, 0, 0|] and is of type [|Int|]
+```
+
+Example 7:
+Iterating over multidimensional arrays.
+
+```qsharp
+let data3 = [
+    |
+        |0, 1, 2|,
+        |3, 4, 5|,
+        |6, 7, 8|
+    |,
+
+    |
+        |9, 10, 11|,
+        |12, 13, 14|,
+        |15, 16, 17|
+    |,
+
+    |
+        |18, 19, 20|,
+        |21, 22, 23|,
+        |24, 25, 26|
+    |
+];
+// data3 is of type [||Int||].
+
+for plane in data {
+    // plane is of type [|Int|]
+    for row in plane {
+        // row is of type [Int]
+        Message($"{row}");
+    }
+
+    Message("");
+}
+// Output
+// ======
+// [0, 1, 2]
+// [3, 4, 5]
+// [6, 7, 8]
+//
+// [9, 10, 11]
+// [12, 13, 14]
+// [15, 16, 17]
+//
+// [18, 19, 20]
+// [21, 22, 23]
+// [24, 25, 26]
+
+// Iteration "on the left" is equivalent to iterating over the leftmost index
+// of a multidimensional array:
+let data2 = [
+    |0, 1|,
+    |2, 3|
+];
+// In particular, the following loops are identical up to that the idxRow
+// variable is not defined in the second loop.
+for idxRow in 0..Fst(Size2(data2)) - 1 {
+    let row = data2[idxRow, ...];
+    // ...
+}
+for row in data2 {
+    // ...
+}
+```
+
+Example 7:
+Implementing array library functions using internal functions.
+
+```qsharp
+namespace Microsoft.Quantum.Arrays {
+    open Microsoft.Quantum.Diagnostics;
+
+    function Transposed2<'T>(array : [|'T|]) : [|'T|] {
+        // Start by deconstructing the input using the internal intrinsic
+        // functions from this proposal; see Implementation below.
+        let data = NDArrayData<'T, [|'T|]>(array);
+        let strides = NDArrayStridesUnsafe(array);
+        let offset = NDArrayOffsetUnsafe(array);
+        let size = NDArraySizeUnsafe(array);
+
+        // Now use the internal AsNDArrayUnsafe function
+        // to reconstruct, but with size and strides
+        // reversed.
+        return AsNDArrayUnsafe<'T, [|'T|]>(data, strides[...-1...], offset, size[...-1...]);
+    }
+
+    function ConstantArray2<'T>(size : (Int, Int), element : 'T) : [|'T|] {
+        Fact(Fst(size) >= 0, "First axis had negative length.");
+        Fact(Snd(size) >= 0, "Second axis had negative length.");
+
+        // Here, we set a stride of zero to store everything as a single
+        // element. Using the copy-and-update operator will require actually
+        // allocating the whole array, but we can start off by "cheating."
+        return AsNDArrayUnsafe<'T, [|'T|]>([element], [0, 0], 0, size);
```

Comment:
@cgranade You are using the AsNDArrayUnsafe in an inconsistent manner; size here is a tuple rather than an array, and the data array needs to have more than one element in general.
In Example 9 below, FoldedAlongAxis2 is not defined. I don't think either of those two examples needs or necessarily should be part of the language proposal; this really is the implementation and we should not have full pieces of the implementation in the proposal. I hence suggest to remove both Example 8 and 9 from the proposal.

Reply:
Sounds good, will drop both, then.